File: GCRTJS/box.js
class Box
{
	constructor(glIn)
	{
		this.gl = glIn;
		this.vaoID = this.gl.createVertexArray();
		this.gl.bindVertexArray(this.aoID);

		this.nVerts = 36;

		this.genPositions();
		this.genNormals();
		this.genUVs();
		this.genTans();

		this.model = mat4.create();
	}

	genPositions()
	{
		var pos = [];

		var hx = 1.0;
	    var hy = 1.0;
	    var hz = 1.0;

	    // Front

	    pos.push(vec3.fromValues(hx, hy, hz));
	    pos.push(vec3.fromValues(hx, -hy, hz));
	    pos.push(vec3.fromValues(hx, -hy, -hz));
	    pos.push(vec3.fromValues(hx, hy, hz));
	    pos.push(vec3.fromValues(hx, -hy, -hz));
	    pos.push(vec3.fromValues(hx, hy, -hz));
	    pos.push(vec3.fromValues(-hx, -hy, hz));
	    pos.push(vec3.fromValues(-hx, hy, hz));
	    pos.push(vec3.fromValues(-hx, hy, -hz));
	    pos.push(vec3.fromValues(-hx, -hy, hz));
	    pos.push(vec3.fromValues(-hx, hy, -hz));
	    pos.push(vec3.fromValues(-hx, -hy, -hz));
	    pos.push(vec3.fromValues(-hx, hy, hz));
	    pos.push(vec3.fromValues(hx, hy, hz));
	    pos.push(vec3.fromValues(hx, hy, -hz));
	    pos.push(vec3.fromValues(-hx, hy, hz));
	    pos.push(vec3.fromValues(hx, hy, -hz));
	    pos.push(vec3.fromValues(-hx, hy, -hz));
	    pos.push(vec3.fromValues(hx, -hy, hz));
	    pos.push(vec3.fromValues(-hx, -hy, hz));
	    pos.push(vec3.fromValues(-hx, -hy, -hz));
	    pos.push(vec3.fromValues(hx, -hy, hz));
	    pos.push(vec3.fromValues(-hx, -hy, -hz));
	    pos.push(vec3.fromValues(hx, -hy, -hz));
	    pos.push(vec3.fromValues(hx, hy, hz));
	    pos.push(vec3.fromValues(-hx, hy, hz));
	    pos.push(vec3.fromValues(-hx, -hy, hz));
	    pos.push(vec3.fromValues(hx, hy, hz));
	    pos.push(vec3.fromValues(-hx, -hy, hz));
	    pos.push(vec3.fromValues(hx, -hy, hz));
	    pos.push(vec3.fromValues(-hx, hy, -hz));
	    pos.push(vec3.fromValues(hx, hy, -hz));
	    pos.push(vec3.fromValues(hx, -hy, -hz));
	    pos.push(vec3.fromValues(-hx, hy, -hz));
	    pos.push(vec3.fromValues(hx, -hy, -hz));
	    pos.push(vec3.fromValues(-hx, -hy, -hz));

	    var pos32 = new Float32Array(pos.length * 3);

	    for(var i = 0; i < pos.length; i++)
	    {
	    	pos32[3 * i] = pos[i].x;
	    	pos32[3 * i + 1] = pos[i].y;
	    	pos32[3 * i + 2] = pos[i].z;
	    }

	    this.posVboID = this.gl.createBuffer();
	    this.gl.bindBuffer(this.gl.ARRAY_BUFFER, this.posVboID);
	    this.gl.bufferData(this.gl.ARRAY_BUFFER, pos32, gl.STATIC_DRAW, 0, pos32.length);

	    this.gl.vertexAttribPointer(0, 3, this.gl.FLOAT, gl.FALSE, 0, 0);
    	this.gl.enableVertexAttribArray(0);
	}

	genNormals()
	{
		var norms = [];

	    norms.push(vec3.fromValues(1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, -1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, -1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, -1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, -1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, -1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, -1.0, 0.0));
	    norms.push(vec3.fromValues(0.0, 0.0, 1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, 1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, 1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, 1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, 1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, 1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, -1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, -1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, -1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, -1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, -1.0));
	    norms.push(vec3.fromValues(0.0, 0.0, -1.0));

	    var norms32 = new Float32Array(norms.length * 3);

	    for(var i = 0; i < norms.length; i++)
	    {
	    	norms32[3 * i]     = norms[i].x;
	    	norms32[3 * i + 1] = norms[i].y;
	    	norms32[3 * i + 2] = norms[i].z;
	    }

	    this.normVboID = this.gl.createBuffer();
	    this.gl.bindBuffer(this.gl.ARRAY_BUFFER, this.normVboID);
	    this.gl.bufferData(this.gl.ARRAY_BUFFER, norms32, this.gl.STATIC_DRAW, 0, norms32.length);

	    this.gl.vertexAttribPointer(1, 3, this.gl.FLOAT, gl.FALSE, 0, 0);
    	this.gl.enableVertexAttribArray(1);
	}

	genUVs()
	{
		var uvs = [];

	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 0.0));
	    uvs.push(vec2.fromValues(0.0, 1.0));
	    uvs.push(vec2.fromValues(1.0, 1.0));

	    var uvs32 = new Float32Array(uvs.length * 2);

	    for(var i = 0; i < uvs.length; i++)
	    {
	    	uvs32[2 * i]     = uvs[i].x;
	    	uvs32[2 * i + 1] = uvs[i].y;
	    }

	    this.normVboID = this.gl.createBuffer();
	    this.gl.bindBuffer(this.gl.ARRAY_BUFFER, this.normVboID);
	    this.gl.bufferData(this.gl.ARRAY_BUFFER, uvs32, this.gl.STATIC_DRAW, 0, uvs32.length);

	    this.gl.vertexAttribPointer(2, 2, this.gl.FLOAT, gl.FALSE, 0, 0);
    	this.gl.enableVertexAttribArray(2);
	}

	genTans()
	{
		var tans = [];

	    tans.push(vec3.fromValues(0.0, 1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, 1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, 1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, 1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, 1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, 1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, -1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, -1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, -1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, -1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, -1.0, 0.0));
	    tans.push(vec3.fromValues(0.0, -1.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));
	    tans.push(vec3.fromValues(-1.0, 0.0, 0.0));

	    var tans32 = new Float32Array(tans.length * 3);

	    for(var i = 0; i < tans.length; i++)
	    {
	    	tans32[3 * i]     = tans[i].x;
	    	tans32[3 * i + 1] = tans[i].y;
	    	tans32[3 * i + 2] = tans[i].z;
	    }

	    this.tanVboID = this.gl.createBuffer();
	    this.gl.bindBuffer(this.gl.ARRAY_BUFFER, this.tanVboID);
	    this.gl.bufferData(this.gl.ARRAY_BUFFER, tans32, this.gl.STATIC_DRAW, 0, tans32.length);

	    this.gl.vertexAttribPointer(3, 3, this.gl.FLOAT, gl.FALSE, 0, 0);
    	this.gl.enableVertexAttribArray(3);
	}

	setModel(modelIn)
	{
		this.model = modelIn;
	}

	get Model()
	{
		return this.model;
	}

	draw()
	{
		this.gl.bindVertexArray(this.vaoID);
		this.gl.drawArrays(this.gl.TRIANGLES, 0, this.nVerts);
	}
}

File: gcrtcommon/src/box.cpp
#include "box.h"

/**
 * GenPositions Generate vertex positions for a box.
 * @param pos Vector to fill with box positions.
 */

void GenPositionsBox(vector<vec3> &pos)
{
    float hx = 1.0;
    float hy = 1.0;
    float hz = 1.0;

    // Front

    pos.push_back(vec3(hx, hy, hz));
    pos.push_back(vec3(hx, -hy, hz));
    pos.push_back(vec3(hx, -hy, -hz));

    pos.push_back(vec3(hx, hy, hz));
    pos.push_back(vec3(hx, -hy, -hz));
    pos.push_back(vec3(hx, hy, -hz));

    // Back

    pos.push_back(vec3(-hx, -hy, hz));
    pos.push_back(vec3(-hx, hy, hz));
    pos.push_back(vec3(-hx, hy, -hz));

    pos.push_back(vec3(-hx, -hy, hz));
    pos.push_back(vec3(-hx, hy, -hz));
    pos.push_back(vec3(-hx, -hy, -hz));

    // Right

    pos.push_back(vec3(-hx, hy, hz));
    pos.push_back(vec3(hx, hy, hz));
    pos.push_back(vec3(hx, hy, -hz));

    pos.push_back(vec3(-hx, hy, hz));
    pos.push_back(vec3(hx, hy, -hz));
    pos.push_back(vec3(-hx, hy, -hz));

    // Left

    pos.push_back(vec3(hx, -hy, hz));
    pos.push_back(vec3(-hx, -hy, hz));
    pos.push_back(vec3(-hx, -hy, -hz));

    pos.push_back(vec3(hx, -hy, hz));
    pos.push_back(vec3(-hx, -hy, -hz));
    pos.push_back(vec3(hx, -hy, -hz));

    // Top

    pos.push_back(vec3(hx, hy, hz));
    pos.push_back(vec3(-hx, hy, hz));
    pos.push_back(vec3(-hx, -hy, hz));

    pos.push_back(vec3(hx, hy, hz));
    pos.push_back(vec3(-hx, -hy, hz));
    pos.push_back(vec3(hx, -hy, hz));

    // Bottom

    pos.push_back(vec3(-hx, hy, -hz));
    pos.push_back(vec3(hx, hy, -hz));
    pos.push_back(vec3(hx, -hy, -hz));

    pos.push_back(vec3(-hx, hy, -hz));
    pos.push_back(vec3(hx, -hy, -hz));
    pos.push_back(vec3(-hx, -hy, -hz));
}

/**
 * GenNormals Get vertex normals for a box.
 * @param norms Vector of normals to fill.
 */

void GenNormalsBox(vector<vec3> &norms)
{
    float hx = 0.5;
    float hy = 0.5;
    float hz = 0.5;

    // Front

    norms.push_back(vec3(1.0, 0.0, 0.0));
    norms.push_back(vec3(1.0, 0.0, 0.0));
    norms.push_back(vec3(1.0, 0.0, 0.0));

    norms.push_back(vec3(1.0, 0.0, 0.0));
    norms.push_back(vec3(1.0, 0.0, 0.0));
    norms.push_back(vec3(1.0, 0.0, 0.0));

    // Back

    norms.push_back(vec3(-1.0, 0.0, 0.0));
    norms.push_back(vec3(-1.0, 0.0, 0.0));
    norms.push_back(vec3(-1.0, 0.0, 0.0));

    norms.push_back(vec3(-1.0, 0.0, 0.0));
    norms.push_back(vec3(-1.0, 0.0, 0.0));
    norms.push_back(vec3(-1.0, 0.0, 0.0));

    // Right

    norms.push_back(vec3(0.0, 1.0, 0.0));
    norms.push_back(vec3(0.0, 1.0, 0.0));
    norms.push_back(vec3(0.0, 1.0, 0.0));

    norms.push_back(vec3(0.0, 1.0, 0.0));
    norms.push_back(vec3(0.0, 1.0, 0.0));
    norms.push_back(vec3(0.0, 1.0, 0.0));

    // Left

    norms.push_back(vec3(0.0, -1.0, 0.0));
    norms.push_back(vec3(0.0, -1.0, 0.0));
    norms.push_back(vec3(0.0, -1.0, 0.0));

    norms.push_back(vec3(0.0, -1.0, 0.0));
    norms.push_back(vec3(0.0, -1.0, 0.0));
    norms.push_back(vec3(0.0, -1.0, 0.0));

    // Top

    norms.push_back(vec3(0.0, 0.0, 1.0));
    norms.push_back(vec3(0.0, 0.0, 1.0));
    norms.push_back(vec3(0.0, 0.0, 1.0));

    norms.push_back(vec3(0.0, 0.0, 1.0));
    norms.push_back(vec3(0.0, 0.0, 1.0));
    norms.push_back(vec3(0.0, 0.0, 1.0));

    // Bottom

    norms.push_back(vec3(0.0, 0.0, -1.0));
    norms.push_back(vec3(0.0, 0.0, -1.0));
    norms.push_back(vec3(0.0, 0.0, -1.0));

    norms.push_back(vec3(0.0, 0.0, -1.0));
    norms.push_back(vec3(0.0, 0.0, -1.0));
    norms.push_back(vec3(0.0, 0.0, -1.0));
}

/**
 * GenUVs Generate vertex UVs for a box.
 * @param uvs Vector of UVs to fill.
 */

void GenUVsBox(vector<vec2> &uvs)
{
    float hx = 0.5;
    float hy = 0.5;
    float hz = 0.5;

    // Front

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));
    uvs.push_back(vec2(1.0f, 1.0f));

    // Back

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));
    uvs.push_back(vec2(1.0f, 1.0f));

    // Right

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));
    uvs.push_back(vec2(1.0f, 1.0f));

    // Left

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));
    uvs.push_back(vec2(1.0f, 1.0f));

    // Top

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));
    uvs.push_back(vec2(1.0f, 1.0f));

    // Bottom

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));

    uvs.push_back(vec2(1.0f, 0.0f));
    uvs.push_back(vec2(0.0f, 1.0f));
    uvs.push_back(vec2(1.0f, 1.0f));
}

/**
 * GenTans Generate vertex tangents for a box.
 * @param tans Vector of tangents to fill.
 */

void GenTansBox(vector<vec3> &tans)
{
    // Front

    tans.push_back(vec3(0.0, 1.0, 0.0));
    tans.push_back(vec3(0.0, 1.0, 0.0));
    tans.push_back(vec3(0.0, 1.0, 0.0));

    tans.push_back(vec3(0.0, 1.0, 0.0));
    tans.push_back(vec3(0.0, 1.0, 0.0));
    tans.push_back(vec3(0.0, 1.0, 0.0));

    // Back

    tans.push_back(vec3(0.0, -1.0, 0.0));
    tans.push_back(vec3(0.0, -1.0, 0.0));
    tans.push_back(vec3(0.0, -1.0, 0.0));

    tans.push_back(vec3(0.0, -1.0, 0.0));
    tans.push_back(vec3(0.0, -1.0, 0.0));
    tans.push_back(vec3(0.0, -1.0, 0.0));

    // Right

    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));

    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));

    // Left

    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));

    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));

    // Top

    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));

    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));
    tans.push_back(vec3(1.0, 0.0, 0.0));

    // Bottom

    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));

    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));
    tans.push_back(vec3(-1.0, 0.0, 0.0));
}